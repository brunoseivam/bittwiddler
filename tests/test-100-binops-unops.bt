main {
    var a_i32 : int32 = 3;
    var b_i32 : int32 = 2;
    var c_i32 : int32 = 0;
    var c_bool : bool = false;

    # Deliberate mixed use of ' and " delimited strings

    emit('Binary operations on integers\n');
    c_i32 = a_i32 + b_i32;
    emit("(int32) {a_i32} + {b_i32} = {c_i32}\n");
    c_i32 = a_i32 - b_i32;
    emit("(int32) {a_i32} - {b_i32} = {c_i32}\n");
    c_i32 = a_i32 * b_i32;
    emit("(int32) {a_i32} * {b_i32} = {c_i32}\n");
    c_i32 = a_i32 / b_i32;
    emit("(int32) {a_i32} / {b_i32} = {c_i32}\n");
    c_i32 = a_i32 % b_i32;
    emit("(int32) {a_i32} % {b_i32} = {c_i32}\n");
    c_i32 = a_i32 << b_i32;
    emit("(int32) {a_i32} << {b_i32} = {c_i32}\n");
    c_i32 = a_i32 >> b_i32;
    emit("(int32) {a_i32} >> {b_i32} = {c_i32}\n");
    c_i32 = a_i32 | b_i32;
    emit("(int32) {a_i32} | {b_i32} = {c_i32}\n");
    c_i32 = a_i32 & b_i32;
    emit("(int32) {a_i32} & {b_i32} = {c_i32}\n");
    c_bool = (a_i32 != 0) or (b_i32 != 0);
    emit("(int32) {a_i32} or {b_i32} = {c_bool}\n");
    c_bool = (a_i32 != 0) and (b_i32 != 0);
    emit("(int32) {a_i32} and {b_i32} = {c_bool}\n");
    c_bool = a_i32 < b_i32;
    emit("(int32) {a_i32} < {b_i32} = {c_bool}\n");
    c_bool = a_i32 > b_i32;
    emit("(int32) {a_i32} > {b_i32} = {c_bool}\n");
    c_bool = a_i32 <= b_i32;
    emit("(int32) {a_i32} <= {b_i32} = {c_bool}\n");
    c_bool = a_i32 >= b_i32;
    emit("(int32) {a_i32} >= {b_i32} = {c_bool}\n");
    c_bool = a_i32 == b_i32;
    emit("(int32) {a_i32} == {b_i32} = {c_bool}\n");
    c_bool = a_i32 != b_i32;
    emit("(int32) {a_i32} != {b_i32} = {c_bool}\n");

    emit('\nUnary operations on integers\n');
    c_i32 = ~a_i32;
    emit("(int32) ~{a_i32} = {c_i32}\n");
    c_bool = not (a_i32!=0);
    emit("(int32) not {a_i32} = {c_bool}\n");
    c_i32 = -a_i32;
    emit('(int32) -{a_i32} = {c_i32}\n');

    var a_f64 : float64 = 3.0;
    var b_f64 : float64 = 2.0;
    var c_f64 : float64 = 0.0;

    emit('\nBinary operations on floats\n');
    c_f64 = a_f64 + b_f64;
    emit("(float64) {a_f64} + {b_f64} = {c_f64}\n");
    c_f64 = a_f64 - b_f64;
    emit("(float64) {a_f64} - {b_f64} = {c_f64}\n");
    c_f64 = a_f64 * b_f64;
    emit("(float64) {a_f64} * {b_f64} = {c_f64}\n");
    c_f64 = a_f64 / b_f64;
    emit("(float64) {a_f64} / {b_f64} = {c_f64}\n");
    c_bool = a_f64 < b_f64;
    emit("(float64) {a_f64} < {b_f64} = {c_bool}\n");
    c_bool = a_f64 > b_f64;
    emit("(float64) {a_f64} > {b_f64} = {c_bool}\n");
    c_bool = a_f64 <= b_f64;
    emit("(float64) {a_f64} <= {b_f64} = {c_bool}\n");
    c_bool = a_f64 >= b_f64;
    emit("(float64) {a_f64} >= {b_f64} = {c_bool}\n");
    c_bool = a_f64 == b_f64;
    emit("(float64) {a_f64} == {b_f64} = {c_bool}\n");
    c_bool = a_f64 != b_f64;
    emit("(float64) {a_f64} != {b_f64} = {c_bool}\n");

    emit('\nUnary operations on floats\n');
    c_f64 = -a_f64;
    emit('(float64) -{a_f64} = {c_f64}\n');
}
