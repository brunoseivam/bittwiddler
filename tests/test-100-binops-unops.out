Binary operations on integers
(int32) 3 + 2 = 5
(int32) 3 - 2 = 1
(int32) 3 * 2 = 6
(int32) 3 / 2 = 1
(int32) 3 % 2 = 1
(int32) 3 << 2 = 12
(int32) 3 >> 2 = 0
(int32) 3 | 2 = 3
(int32) 3 & 2 = 2
(int32) 3 or 2 = 1
(int32) 3 and 2 = 1
(int32) 3 < 2 = 0
(int32) 3 > 2 = 1
(int32) 3 <= 2 = 0
(int32) 3 >= 2 = 1
(int32) 3 == 2 = 0
(int32) 3 != 2 = 1

Unary operations on integers
(int32) ~3 = -4
(int32) not 3 = 0
(int32) -3 = -3

Binary operations on floats
(float64) 3.000000 + 2.000000 = 5.000000
(float64) 3.000000 - 2.000000 = 1.000000
(float64) 3.000000 * 2.000000 = 6.000000
(float64) 3.000000 / 2.000000 = 1.500000
(float64) 3.000000 < 2.000000 = 0
(float64) 3.000000 > 2.000000 = 1
(float64) 3.000000 <= 2.000000 = 0
(float64) 3.000000 >= 2.000000 = 1
(float64) 3.000000 == 2.000000 = 0
(float64) 3.000000 != 2.000000 = 1

Unary operations on floats
(float64) -3.000000 = -3.000000
